examples/tutorials/mnist_pytorch/idle_gpu_watcher.py

```diff
@@ -0,0 +1,178 @@
+        self._full_window = sample_freq * num_samples
+        self._min_threshhold = min_threshhold_percentage
+        self._delay_samples = delay_samples
+        self._debug = debug
+
+        header_regstr = {
+            "uuid": r"(?P<uuid>[a-zA-Z0-9-]+)",
+            "utilization.gpu": r"(?P<gpu_util>[0-9]+) %"
+        }
+        line_regstr = ", ".join(header_regstr.values())
+        query_gpu = ",".join(header_regstr.keys())
+
+        self._nvida_smi_cmd = ["nvidia-smi", "--format=csv,noheader", f"--query-gpu={query_gpu}"]
+        self._line_regex = re.compile(line_regstr)
+        self._gpu_util_samples = {}  # gpu_uuid: <circular queue of samples>
+
+    def _get_nvidia_smi(self):
+        # nvidia-smi --help-query-gpu: "utilization.gpu":
+        # Percent of time over the past sample period during which one or more kernels was
+        # executing on the GPU. The sample period may be between 1 second and 1/6 second
+        # depending on the product.
+        output = {}
+
+        csv_output = subprocess.check_output(self._nvida_smi_cmd).decode()
+        for line in csv_output.splitlines():
+            match = re.match(self._line_regex, line)
+            if match is None:
+                raise RuntimeError(f"Unexpected output format: {line}")
+
+            data = match.groupdict()
+            output[data["uuid"]] = int(data["gpu_util"])
+        return output
+
+    def _check_idle(self):
+        gpu_utils = self._get_nvidia_smi()
+
+        for gpu_uuid, gpu_util in gpu_utils.items():
+            if self._gpu_util_samples.get(gpu_uuid) is None:
+                self._gpu_util_samples[gpu_uuid] = deque()
+
+            self._gpu_util_samples[gpu_uuid].append(gpu_util)
+
+            if len(self._gpu_util_samples[gpu_uuid]) < self._num_samples:
+                continue
+
+            if all(s < self._min_threshhold for s in self._gpu_util_samples[gpu_uuid]):
+                sys.stderr.write(
+                    f"IDLE_GPU_WATCHER: {gpu_uuid} usage below threshhold {self._min_threshhold}% "
+                    f"for last {self._sample_freq * self._num_samples} seconds. Killing trial.\n")
+                sys.exit(1)
+            if self._debug:
+                print(f"DEBUG: {gpu_uuid} usage {gpu_util}% ")
+
+            self._gpu_util_samples[gpu_uuid].popleft()
+
+    def watch_process(self, args):
+        p = subprocess.Popen(args)
+        print("IDLE_GPU_WATHCER: "
+              f"Starting idle GPU watcher: min_threshhold={self._min_threshhold}%, "
+              f"sample_freq={self._sample_freq}, num_samples={self._num_samples}, "
+              f"delay_samples={self._delay_samples}, window_size={self._full_window}")
+        try:
+            while True:
+                if p.poll() is None:
+                    time.sleep(self._sample_freq)
+
+                    if self._delay_samples > 0:
+                        self._delay_samples -= 1
+                        continue
```

Comment:
I would personally recommend that you write the delay differently.

As-is, you'll decrement the delay to zero, then wait for the sample deque to fill up, then start working.

It might be more intuitive to the user to have the `delay_samples` be an absolute setting, rather than something that combines with the windows size (effectively).  So you would increment a num_samples, and after you've filled your sample deque, you would do an additional check that num_samples was > delay_samples, or you'd skip the `if all(...)` check.

Not sure it matters a whole lot though.